valhalla/baldr/graphtile.h

```diff
@@ -829,11 +815,52 @@ class GraphTile : public boost::intrusive_ref_counter<GraphTile, GraphTileRefCou
   /** Decrompresses tile bytes into the internal graphtile byte buffer
    * @param  graphid     the id of the tile to be decompressed
    * @param  compressed  the compressed bytes
-   * @return a pointer to a graphtile if it  has been successfully initialized with
-   *         the uncompressed data, or nullptr
-   */
-  static graph_tile_ptr DecompressTile(const GraphId& graphid, const std::vector<char>& compressed);
+   * @return a pointer to the managed memory of the graph tile if it has been successfully
+   *         decompressed otherwise nullptr
+   */
+  static std::unique_ptr<const GraphMemory> DecompressTile(const GraphId& graphid,
+                                                           const std::vector<char>& compressed);
+
+  // we encapsulate tile pointers inside of boost::intrusive_ptr which requires reference counting
+  // the default reference counting provided allows for both thread safe and unsafe modes of operation
+  // but.. its templated which makes it impossible to decide at runtime which you need
+  // global_synchronized_cache requires thread safe tiles when they manage their own lifetime
+  // this tile cache tells graphtile to use an atomic counter (which is about %5 more expensive)
+  // all other tile caches are used by individual threads so they dont need to do atomic operations
+  std::unique_ptr<ref_counter> ref_counter_;
```

Comment:
this is getting close. all that is left to do is to change graphreader to properly construct safe or unsafe variants based on the the t ype of tile cache we have. @dnesbitt61 ill ping you for testing as soon as i get that in